### ISSUE TITLE
FastxFile truncates gzip compressed fastq files

### ISSUE THREAD

**Issue by davidtsao** [pysam-developers/pysam#738]
I've noticed that using pysam.FastxFile to iterate over a bgzip'd fastq file only returns a subset of fastq records. The expected result is that all records are iterated over. 

Example:

example.fastq.gz contains 1,009,470 reads. 

When I iterate over the compressed fastq file using pysam, I only get 26,107 reads instead of the expected 1,009,470 reads. 

```
In [1]: import pysam

In [2]: len([x for x in pysam.FastxFile('example.fastq.gz')])
Out[2]: 26107
```

I've traced the issue back to bgzip not decompressing all the contents of the fastq file.

For example, I get the expected ~4M lines using gunzip to decompress:
```
$ bgzip -c -d example.fastq.gz | wc -l 
 4037880
```

But using bgzip to decompress only returns ~104k lines, which is the same result obtained from pysam (26107 * 4 = 104428):

```
$ bgzip -c -d example.fastq.gz | wc -l
  104428
```

**Comment by AndreasHeger** [comment]
If you could provide a PR for docs, that would be great, thanks!
I agree that this is a serious issue and leave it open - if I have time will look into this to see if the gzip iterator is recoverable from a bogus EOF.

**Comment by cgjosephlee** [comment]
Any flow-ups of this issue?

**Comment by jmarshall** [comment]
Without an example file, it is unclear whether this might be early stopping due to extra BGZF trailer blocks (samtools/htslib#45) or early stopping due to multiple GZIP members in a plain GZIP (non-bgzipped) file (samtools/htslib#742) or due to another reason.

The former has been fixed since HTSlib 1.4 so it is probably not that.

The latter has been fixed on HTSlib **develop** but after the 1.9 release. So if this problem is that, it will be fixed in an upcoming release.

**Comment by cgjosephlee** [comment]
I guess I'm in the second problem, looking forward to it. Thank you.

**Comment by zmunro** [comment]
Is a fix on the way for this? I am still experiencing this issue.

**Comment by jmarshall** [comment]
There is now an HTSlib release containing the fix for the second problem noted in https://github.com/pysam-developers/pysam/issues/738#issuecomment-487958180, so when pysam's HTSlib is updated that problem will go away.

If you would like to provide an example file exhibiting the issue you're experiencing, we can investigate whether the problem is one of those two or a different problem still needing fixing.

**Comment by zmunro** [comment]
@jmarshall [Here is a link](https://drive.google.com/file/d/16TDoy9hibyYCSL7NH4Z9HNV9wsfLsq_g/view?usp=sharing) to a fastqgz file that I have made publicly downloadable from my google drive. This is the file that I was having issues with. Here is the code I was running on it:
```
num_reads = 0
with pysam.FastxFile(path_to_fastqgz) as fastqgz_file:
    for entry in fastqgz_file:
        num_reads +=1
print("num reads: " + str(num_reads))
```
And the output is 22 with the fastqgz file I provided. However if you decompress the fastqgz file and instead run the code on a fastq, the output is 154470.
I am using pysam==0.15.3, and python3.7.4

**Comment by bioinformed** [comment]
We are working on a (large) update to pysam to support htslib 1.10.x.
There is no specific ETA yet.

-Kevin

**Comment by jmarshall** [comment]
@zmunro: I can confirm that your file is one that is fixed by samtools/htslib#744 so will work in an upcoming pysam release incorporating the recent HTSlib release.

**Comment by luyang93** [comment]
https://github.com/samtools/htslib/pull/744#issuecomment-635778154

**Comment by jmarshall** [comment]
This has been fixed since pysam 0.16.0 (when used with HTSlib 1.10 or later).
